rust/uefi/linux-bootloader/src/measure.rs

```diff
@@ -26,43 +27,95 @@ pub fn measure_image(system_table: &SystemTable<Boot>, image: PeInMemory) -> uef
-                    info!("Measuring section `{}`...", section_name);
-                    if tpm_log_event_ascii(
-                        boot_services,
-                        TPM_PCR_INDEX_KERNEL_IMAGE,
-                        data,
-                        section_name,
-                    )? {
-                        measurements += 1;
-                    }
-                }
+
+    // Match behaviour of systemd-stub (see src/boot/efi/stub.c in systemd)
+    // The encoding as well as the ordering of measurements is critical.
+    //
+    // References:
+    //
+    // "TPM2 PCR Measurements Made by systemd", https://systemd.io/TPM2_PCR_MEASUREMENTS/
+    //   Section: PCR Measurements Made by systemd-stub (UEFI)
+    //   - PCR 11, EV_IPL, “PE Section Name”
+    //   - PCR 11, EV_IPL, “PE Section Data”
+    //
+    // Unified Kernel Image (UKI) specification, UAPI Group,
+    // https://uapi-group.org/specifications/specs/unified_kernel_image/#uki-tpm-pcr-measurements
+    //
+    // Citing from "UKI TPM PCR Measurements":
+    //   On systems with a Trusted Platform Module (TPM) the UEFI boot stub shall measure the sections listed above,
+    //   starting from the .linux section, in the order as listed (which should be considered the canonical order).
+    //   The .pcrsig section is not measured.
+    //
+    //   For each section two measurements shall be made into PCR 11 with the event code EV_IPL:
+    //    - The section name in ASCII (including one trailing NUL byte)
+    //    - The (binary) section contents
+    //
+    //   The above should be repeated for every section defined above, so that the measurements are interleaved:
+    //   section name followed by section data, followed by the next section name and its section data, and so on.
+
+    // NOTE: The order of measurements is important, so the use of BTreeMap is intentional here.
+    let ordered_sections: BTreeMap<_, _> = pe
+        .sections
+        .iter()
+        .filter_map(|section| {
+            let section_name = section.name().ok()?;
+            let unified_section = UnifiedSection::try_from(section_name).ok()?;
+            if !unified_section.should_be_measured() {
+                return None;
             }
+            Some((unified_section, section))
+        })
+        .collect();
+
+    for (_unified_section, section) in ordered_sections {
+        // Unwrapping is safe: We already read the section name above
+        let section_name = section.name().unwrap();
+
+        info!("Measuring section `{}`...", section_name);
+
+        // First measure the section name itself
+        // This needs to be an UTF-8 encoded string with a trailing null byte
+        //
+        // As per reference:
+        // "Measured hash covers the PE section name in ASCII (including a trailing NUL byte!)."
+        let section_name_cs_utf8 = CString::new(section_name).unwrap();
+
+        if tpm_log_event_ascii(
```

Comment:
@RaitoBezarius Do you know why this is called `_ascii`?

Reply:
The name probably leans on the systemd equivalent, which is also called `tpm_log_event_ascii`:
https://github.com/systemd/systemd/blob/fdd4263cac327a3bc447560dbd46713524046099/src/boot/efi/measure.c#L281-L288

Reply:
Yeah, it's just a TPM log event wrapper that applies preprocessing to the description so that userspace can read it.